neural_compressor/compression/pruner/pruners.py

```diff
@@ -901,3 +904,255 @@ def print_progressive_sparsity(self):
         """Output the progressive sparsity."""
         cur_sp = self.pattern.get_sparsity_ratio_progressive(self.progressive_masks)
         logger.info("Step: {} -> Current progressive sparsity: {}".format(self.global_step, cur_sp))
+
+@register_pruner('sparsegpt')
+class SparseGPTPruner(BasePruner):
+    """Pruning Pruner.
+    The SparseGPT pruner_class is derived from BasePruner.
+    SparseGPT Pruner only supports one-shot pruning (same effect as fast retraining free).
+    Please refer to SparseGPT: Massive Language Models Can be Accurately Pruned in One-Shot
+        (https://arxiv.org/abs/2301.00774)
+
+    1. Defines pruning functions called at step begin/end, before/after optimize and epoch begin/end.
+    2. Defines the pruning criterion and fixed weight parameters.
+    3. Get the pruned masks of different layers.
+
+    Args:
+        modules: A dict {"module_name": Tensor} that stores the pruning modules' weights.
+        config: A config dict object that contains the pruner information.
+
+    Attributes:
+        pattern: A Pattern object that defines pruning weights' arrangements within space.
+        criterion: A Criterion Object that defines which weights are to be pruned.
+        scheduler: A Scheduler object that defines how the model's sparsity changes as training/pruning proceeds.
+        reg: A Reg object that defines regulization terms.
+    """
+    def __init__(self, config, modules):
+        """Initialize."""
+        super(SparseGPTPruner, self).__init__(config, modules)
+
+    def _init(self):
+        """Initialize."""
+        self.pattern = get_pattern(self.config, self.modules)
+        self.scheduler = get_scheduler(self.config)
+        logger.warning("SparseGPT pruner fixed the weights, please DO NOT turn on gradient update.")
+
+    def on_train_begin(self, dataloader=None, model=None):
+        """Implement at the beginning of training phase."""
+        self.fasterprune_masks(dataloader, model)
+    
+    def _get_target_layer(self, module, layers=[nn.Conv2d, nn.Linear], name=''):
+        if type(module) in layers:
+            return {name: module}
+        res = {}
+        for name1, child in module.named_children():
+            res.update(self._get_target_layer(
+                child, layers=layers, name=name + '.' + name1 if name != '' else name1
+            ))
+        return res
+
+
+    def fasterprune_masks(self, dataloader, model):
+        ################## preparing for pruning ##################
+        # get device
+        dev = model.device
+        
+        # get layers
+        layers = model.model.decoder.layers
+
+        model.seqlen = model.config.max_position_embeddings
+        
```

Comment:
ok, the logic is correct without bugs till now, but it may need further reconstruction for adaptation to our framework.

Reply:
since sparsegpt is a layer-wise post-training method without the need of gradients like the retraining-free pruner, I put all the logic to the on_train_begin method of the pruner class, which may need further splitting to a new pattern class and criterion class (need discussion with @WeiweiZhang1 )